wiki/python/DEVELOPER_GUIDE.md

```diff
@@ -47,31 +47,110 @@ The examples in the documentation are checked through doctest. To skip the check
-
-Some tests are run against inputs generated with [hypothesis](https://hypothesis.works/). See the `cuml/testing/strategies.py` module for custom strategies that can be used to test cuml estimators with diverse inputs. For example, use the `regression_datasets()` strategy to test random regression problems.
-
-When using hypothesis for testing, you must include at least one explicit example using the `@example` decorator alongside any `@given` strategies. This ensures that:
-1. Every test has at least one deterministic test case that always runs
-2. Critical edge cases are documented and tested consistently
-3. Test failures can be reproduced reliably
-
-Note: While the explicit examples will always run in CI, the hypothesis-generated test cases (from `@given` strategies) only run during nightly testing by default. This ensures fast CI runs while still maintaining thorough testing coverage.
-
-Example of a valid hypothesis test:
-```python
-@example(dtype=np.float32, sparse_input=False)  # baseline case, runs as part of PR CI
-@example(dtype=np.float64, sparse_input=True)   # edge case, runs as part of PR CI
-@given(
-    dtype=st.sampled_from((np.float32, np.float64)),
-    sparse_input=st.booleans()
-)  # strategy-based cases, only runs during nightly tests
-def test_my_estimator(dtype, sparse_input):
-    # Test implementation
-    pass
+We use [pytest](https://docs.pytest.org/en/latest/) for writing and running tests. To see existing examples, refer to any of the `test_*.py` files in the folder `cuml/tests`.
+
+### Test Organization
+- Keep all tests for a single estimator in one file, with exceptions for:
+  - Performance testing/benchmarking
+  - Generic estimator checks (e.g., `test_base.py`)
+- Use small, focused datasets for correctness testing
+- Only parametrize scale when it triggers alternate code paths
+
+### Test Input Generation
+We support three main approaches for test input generation:
+
+1. **Fixtures** (`@pytest.fixture`):
+   - For shared setup/teardown code and resources
+   - Examples: random seeds, clients, loading test datasets
+   ```python
+   @pytest.fixture(scope="module")
+   def random_state():
+       return 42
+   ```
+
+2. **Parametrization** (`@pytest.mark.parametrize`):
+   - For testing specific input combinations
+   - Good for hyperparameters and configurations
+   ```python
+   @pytest.mark.parametrize("solver", ["svd", "eig"])
+   def test_estimator(solver):
+       pass
+   ```
+
+3. **Hypothesis** (`@given`):
+   - For property-based testing with random inputs
+   - Must include at least one `@example` for deterministic testing
+   - Preferred for dataset generation
+   ```python
+   @example(dataset=small_regression_dataset(np.float32))
+   @given(dataset=standard_regression_datasets())
+   def test_estimator(dataset):
+       pass
+   ```
+
+### Test Parameter Levels
+
+You can mark test parameters for different scales with (`unit_param`, `quality_param`, and `stress_param`).
```

Comment:
This might not be a bad time to rething the names of these three or at least define them better. What is the difference between `unit` and `quality`? Perhaps we should use this opportunity to distinguish between tests that we only want to run in nightly runs?

Reply:
Not only are they somewhat purely defined, I'm not fully convinced that we need them at all.

Reply:
I think I agree with Simon. At least I am a bit puzzled what this is for/about. Tests are for testing correctness and benchmarks are for benchmarking. And pytest is not a great tool for writing benchmarks, something like [asv](https://asv.readthedocs.io/en/stable/index.html) is probably more useful for that

Reply:
+1 on removing these fully. Tests should be checking code behavior, and unless logical behavior changes with scale (as noted above and in the linked issue), then what we're doing here is more performance testing which would be better handled by other tools & specific performance tests. I'm very against hiding performance tests in with other tests.

Agree this can be done later, but maybe we want to add a note here dissuading adding more of these (or remove documenting them here/refer to them as legacy code somehow)?